src/query/functions/utils/heap.go

```diff
@@ -0,0 +1,188 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package utils
+
+import (
+	"container/heap"
+)
+
+// ValueIndexPair is a pair of float value and index at which it exists
+type ValueIndexPair struct {
+	Val   float64
+	Index int
+}
+
+type lessFn func(ValueIndexPair, ValueIndexPair) bool
+
+func maxHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val < j.Val
+}
+
+func minHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val > j.Val
+}
+
+// FloatHeap is a heap that can be given a maximum size
+type FloatHeap struct {
+	isMaxHeap bool
+	capacity  int
+	floatHeap *floatHeap
+}
+
+// NewFloatHeap builds a new FloatHeap based on first parameter
+// and a capacity given by second parameter. Zero and negative
+// values for maxSize provide an unbounded FloatHeap
+func NewFloatHeap(isMaxHeap bool, capacity int) FloatHeap {
+	var less lessFn
+	if isMaxHeap {
+		less = maxHeapLess
+	} else {
+		less = minHeapLess
+	}
+
+	floatHeap := &floatHeap{
+		heap: make([]ValueIndexPair, 0, capacity),
+		less: less,
+	}
+
+	heap.Init(floatHeap)
+	return FloatHeap{
+		isMaxHeap: isMaxHeap,
+		capacity:  capacity,
+		floatHeap: floatHeap,
+	}
+}
+
+// Push pushes a value and index pair to the heap
+func (fh FloatHeap) Push(value float64, index int) {
+	h := fh.floatHeap
+	// If capacity is zero or negative, allow infinite size heap
+	if fh.capacity > 0 {
+		// At max size, drop or replace incoming value.
+		// Otherwise, continue as normal
+		if len(h.heap) >= fh.capacity {
+			peek := h.heap[0]
+			// Compare incoming value with current top of heap.
+			// Decide if to replace the current top, or to drop incoming
+			// value as appropriate for this min/max heap
+			//
+			// If values are equal, do not add incoming value regardless
+			//
+			// NB(arnikola): unfortunately, can't just replace first
+			// element as it may not respect internal order. Need to
+			// run heap.Fix() to rectify this
+			if fh.isMaxHeap && value > peek.Val ||
```

Comment:
I don't think we should replicate the heap logic here. A better way would be to have a simple push and pop operation. After doing the push, check the length of the heap. If you're at max, then pop.

Reply:
This is primarily to avoid putting in values we're going to instantly pop off anyway in the `push + pop until max` approach